File: src/schematic/primitives.rs
/// Primitives for drawing circuit schematic elements.

/// A position in 2D space (in egui canvas coordinates).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Pos {
    pub x: f32,
    pub y: f32,
}

impl Pos {
    pub fn new(x: f32, y: f32) -> Self {
        Self { x, y }
    }
}

/// Types of schematic elements that can be drawn.
#[derive(Debug, Clone)]
pub enum SchematicElement {
    /// Voltage source (circle with +/-)
    Source {
        pos: Pos,
        label: String,
        value: String,
    },
    /// Inductor (curved line / zigzag)
    Inductor {
        pos: Pos,
        label: String,
        value: String,
    },
    /// Capacitor (two parallel plates)
    Capacitor {
        pos: Pos,
        label: String,
        value: String,
    },
    /// Diode (triangle + bar)
    Diode {
        pos: Pos,
        label: String,
    },
    /// Switch / MOSFET
    Switch {
        pos: Pos,
        label: String,
    },
    /// Load resistor
    Resistor {
        pos: Pos,
        label: String,
        value: String,
    },
    /// Wire connection
    Wire {
        from: Pos,
        to: Pos,
    },
    /// Connection node (dot) — reserved for future use
    #[allow(dead_code)]
    Node {
        pos: Pos,
        label: String,
    },
    /// Ground symbol
    Ground {
        pos: Pos,
    },
    /// Text label at position
    Label {
        pos: Pos,
        text: String,
    },
}

File: tupan/schematic/primitives.py
"""Primitives for drawing circuit schematic elements.

Mirrors src/schematic/primitives.rs.
"""

from dataclasses import dataclass
from enum import Enum, auto


@dataclass
class Pos:
    """A position in 2D space."""
    x: float
    y: float


class SchematicElement:
    """Base class for schematic elements."""
    pass


@dataclass
class Source(SchematicElement):
    """Voltage source (circle with +/-)."""
    pos: Pos
    label: str = ""
    value: str = ""


@dataclass
class Inductor(SchematicElement):
    """Inductor (curved line / zigzag)."""
    pos: Pos
    label: str = ""
    value: str = ""


@dataclass
class Capacitor(SchematicElement):
    """Capacitor (two parallel plates)."""
    pos: Pos
    label: str = ""
    value: str = ""


@dataclass
class Diode(SchematicElement):
    """Diode (triangle + bar)."""
    pos: Pos
    label: str = ""


@dataclass
class Switch(SchematicElement):
    """Switch / MOSFET."""
    pos: Pos
    label: str = ""


@dataclass
class Resistor(SchematicElement):
    """Load resistor."""
    pos: Pos
    label: str = ""
    value: str = ""


@dataclass
class Wire(SchematicElement):
    """Wire connection."""
    from_pos: Pos
    to_pos: Pos


@dataclass
class Node_(SchematicElement):  # Node is a Python built-in, use Node_
    """Connection node (dot)."""
    pos: Pos
    label: str = ""


@dataclass
class Ground(SchematicElement):
    """Ground symbol."""
    pos: Pos


@dataclass
class Label(SchematicElement):
    """Text label at position."""
    pos: Pos
    text: str = ""
